Modify rule S3001: Adapt to LaYC (#2486)
SonarSource/rspec: rules/S3001/javascript/rule.adoc

== Why is this an issue?

> The `delete` operator is used to remove a property from an object. It only affects https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn[own] properties. There are two valid forms:

Reviewer: ```suggestion
The `delete` operator is used to remove a property from an object. It only affects its https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwn[own] properties. There are two valid ways to remove a property:
```

> * `delete object.property`
> * `delete object[property]`

Reviewer: ```suggestion
* Using the dot notation: `delete object.property`
* Using the bracket notation`delete object[property]`
```

`delete` will throw a `TypeError` in strict mode if the property is a non-configurable property.

`delete identifier` may work if `identifier` is a *configurable* property of the global object. For `identifier` to be *configurable*, it should have been declared directly as a `globalThis` property (`globalThis.identifier = 1`). This form is not common practice and should be avoided. Use `delete globalThis.identifier` instead if needed.

Aside from that case, deleting variables, including function parameters, never works:

* Variables declared with `var` are *non-configurable* properties in the global object, except in CommonJS and ECMAScript modules, where top-level variable declarations are scoped to the module. 
* Variables declared with `let` or `const` are not attached to any object.

> [source,javascript]

Reviewer: Just asking, do we need diff-tag ?

Author: i didn't add them as the code seems unrelated and all of it will be colored

Reviewer: 👍

----

> var x  = 1;

Reviewer: ```suggestion
var x = 1;
```

  ...
  }
};

> delete obj.x;

Reviewer: Can you add a space between `:` and `1` above, meaning:

```javascript
...
x: 1,
...
```

Just trying to be consistent here.
